README.md
# upax_py

This is a Python 3 implementation of the
[Upax](https://jddixon.github.io/xlattice/upax.html)
distributed file system.

## Major System Components

A Upax implementation consists of at least

* a [fault-tolerant log](https://jddixon.github.io/xlattice/ftLog.html)
* a [Upax server](https://jddixon.github.io/xlattice/upaxServer.html)
* a [Upax client](https://jddixon.github.io/xlattice/upaxClient.html) and optionally
* a [Upax mirror](https://jddixon.github.io/xlattice/upaxMirror.html)


Servers involved use a
[consensus algorithm](https://en.wikipedia.org/wiki/Consensus_%28computer_science%29)
to ensure that their pictures of the state of the distributed file system
are consistent.

## Utilities

These include

* importUDir
* upaxBulkPoster
* upaxUpdateNodeID

### importUDir

This utility imports all of the files in a content-keyed U directory into
the Upax distribued file system.

    usage: importUDir [-h] [-1] [-d DESTDIR] [-j] [-s SRCDIR] [-t] [-T] [-V] [-v]

    optional arguments:
      -h, --help            show this help message and exit
      -1, --usingSHA1       use SHA1, and so 160-bit keys
      -d DESTDIR, --destDir DESTDIR
                            base name of destination U/ directory
      -j, --justShow        show args and exit
      -s SRCDIR, --srcDir SRCDIR
                            base name of source U/ directory
      -t, --showTimestamp   show run timestamp
      -T, --testing         test run - write to ./testU
      -V, --showVersion     show version number and date
      -v, --verbose         talk a lot

### upaxBulkPoster

Given a flat directory, imports any constituent files into the Upax
distributd file system.

    usage: post new files in a directory into Upax [-h] [-1] [-e] [-i INDIR] [-H]
                                                   [-j] [-N] [-t] [-u UDIR] [-T]
                                                   [-V] [-v] [-z]

    optional arguments:
      -h, --help            show this help message and exit
      -1, --usingSHA1       use SHA1, and so 160-bit keys
      -e, --ec2Host         set if machine is in EC2
      -i INDIR, --inDir INDIR
                            path to input directory (forced to ./ testIn if
                            testing)
      -H, --hostmaster      set if machine runs bindMgr
      -j, --justShow        show args and exit
      -N, --nameserver      set if machine is a name server and so runs
                            bindLocalMgr
      -t, --showTimestamp   show run timestamp
      -u UDIR, --uDir UDIR  base name of U/ directory, default to Upax
      -T, --testing         test run - write to ./testU
      -V, --showVersion     show version number and date
      -v, --verbose         talk a lot
      -z, --noChanges       don't actually write anything to disk

### upaxUpdateNodeID

The content-keyed store on a Upax system needs a unique ID, its **nodeID**.
This utility writes a randomly chosen 20- or 32-byte value into a file
called *nodeID** at the top level of uDir, so at `uDir/nodeID`.  This is
written as a 40- or 64-character hexadecimal value.

    usage: post new files in a directory into Upax [-h] [-1] [-e] [-i INDIR] [-H]
                                                   [-j] [-N] [-t] [-u UDIR] [-T]
                                                   [-V] [-v] [-z]

    optional arguments:
      -h, --help            show this help message and exit
      -1, --usingSHA1       use SHA1, and so 160-bit keys
      -e, --ec2Host         set if machine is in EC2
      -i INDIR, --inDir INDIR
                            path to .ssh/ (forced to ./ testIn if testing)
      -H, --hostmaster      set if machine runs bindMgr
      -j, --justShow        show args and exit
      -N, --nameserver      set if machine is a name server and so runs
                            bindLocalMgr
      -t, --showTimestamp   show run timestamp
      -u UDIR, --uDir UDIR  base name of U/ directory, default to Upax
      -T, --testing         test run - write to ./testU
      -V, --showVersion     show version number and date
      -v, --verbose         talk a lot
      -z, --noChanges       don't actually write anything to disk

## Dependencies

*   **xlattice_py**, the collection of Python3 XLattice supporrt classes

## Project Status

Incompletely specified, partially implemented.

## On-line Documentation

More information on the **upax_py** project can be found
[here](https://jddixon.github.io/upax_py)


CHANGES
~/dev/py/upax/CHANGES


v0.10.2
    2018-02-28
        * fix setup.py, tox config files                            SLOC 3248
v0.10.1
    2018-02-15
        * import from xlu rather than xlattice.u                    SLOC 3251
v0.10.0
    2018-01-31
        * add blake2b support                                       SLOC 3250
v0.9.8
    2017-12-11
        * general cleanup; some pylinting; minor bug fixes          SLOC 3221
v0.9.7
    2017-09-28
        * add projConfig.toml                                       SLOC 3228
v0.9.6
    2017-09-06
        * add tox.ini, *requirements.txt; pep8                      SLOC 3191
v0.9.5
    2017-07-28
        * rewrite setup.py, MANIFEST.in; minor edits to cfg files   SLOC 2996
v0.9.4
    2017-06-23
        * restructure for pytest; drop zz_mistune_test              SLOC 2745
v0.9.3
    2017-05-05
        * drop dist/* from repos; plus minor edit                   SLOC 3187
v0.9.2
    2017-01-28
        * add dump_options to simplify utilities                    SLOC 3009
v0.9.1
    2017-01-19
        * restructure to get rid of circular import(s)              SLOC 3038
v0.9.0
    2017-01-17
        * drop redundant code files 
        * adapt to new xlattice API (HashTypes etc)                 SLOC 3023
v0.8.5
    2016-11-26
        * further pyinting with minor bugfixes                      SLOC 3223
v0.8.4
    2016-11-18
        * replace most uses of listdir with scandir; more PEP 8     SLOC 3063
v0.8.3
    2016-11-10
        * more PEP 8, more minor bug fixes                          SLOC 3031
v0.8.2
    2016-11-09
        * more PEP 8 plus minor bug fixes                           SLOC 2954
v0.8.1
    2016-11-03
        * update web pages; minor bug fixes; more PEP 8             SLOC 2930
v0.8.0
    2016-11-02
        * rename vars per PEP 8                                     SLOC 2917
v0.7.3
    2016-10-17
        * add pylint, coverage; rename files per PEP 8              SLOC 2911
v0.7.2
    2016-09-13
        * fixed FIX MEs, add -1/2/3 support                         SLOC 2799
v0.7.1
    2016-09-05
        * fix various typos                                         SLOC 2778
v0.7.0
    2016-09-02
        * usingSHA1 -> usingSHA                                     SLOC 2770
v0.6.9
    2016-08-27
        * fix setup.py, license                                     SLOC 2747
v0.6.8
    2016-08-16
        * update docs, referencing xlattice/upax                    SLOC 2711
v0.6.7
    2016-06-29
        * add ghpDoc/{css,img}/                                     SLOC 2697
v0.6.6
    2016-06-16
        * correct bug in putData()                                  SLOC 2261
v0.6.5
    2016-06-14
        * replaced separate tests for usingSHA1 True and False
            with combined tests (dropped several files)
        * fixed bugs relating to usingSHA1                          SLOC 2256
v0.6.4
    2016-06-01
        * add bkp2U, other bash scripts; update .dvcz               SLOC 2607
v0.6.3
    2016-05-11
        * clean up initial commit to github                         SLOC 2557
v0.6.2
    2016-05-10
        * add .dvcz, projCfg*; correct use of xlattice.u256         SLOC 2545
        * actually finish conversion to Python3                     SLOC 2557
v0.6.1
    2015-05-15
        * convert to Python3
    2015-08-10
        * complete the conversion, do some debugging                SLOC 2848
v0.6.0
    2013-01-17
        * dropped redundant usingSHA1 field from LogEntry           SLOC 2265
        * checkUConsistency core in upax.consistency.check          SLOC 2270
        * added property entries to ftlog.Log                       SLOC 2273
        * addEntry now returns LogEntry object                      SLOC 2303
        * checkUConsistency --repair works                          SLOC 2324

    2013-01-16
        * added upax.walker and testUWalker.py                      SLOC 2315
        * checkUConsistency checks content keys but not L           SLOC 2244
            - about 6s for 122K elements
v0.5.1
    2012-10-30
        * copied in, fixed upax3 tests; they succeed                SLOC 2005
            - so we now have test*1.py and test*3.py
v0.5.0
    2012-10-30
        * all SHA1-based tests succeed (also need  to check
            importUDir, upaxBulkPoster, upaxUpdateNodeID)           SLOC 1421
    2012-10-29
        * upax/node.py identical between upax and upax3
v0.4.4
    2012-10-13
        * testUpaxNode.py                                           SLOC 1291
v0.4.3
    2012-10-09
        * added upax/node.py                                        SLOC 1218 
v0.4.2
    2012-09-25
        * corrected typo in upaxBulkPoster                          SLOC 1175
v0.4.1
    2012-07-30
        * fixed bug in Importer.importLeafDir (2-tuple returned 
            from u.put() interpreted as content hash)
        * ran importUDir -s /var/Ue.old -d /var/Ue on PA
v0.4.0
    2012-07-30
        * added importUDir script with core logic in package,
            plus testImporUDir.py; test succeeds after fixes
        * manual check of importUDir using data generated in 
            tests also succeeds after some minor fixes
v0.3.7
    2012-07-26
        * added ATEXT and AT_FREE to define acceptable paths for 
            log entries, specifically to allow RFC2822 message 
            IDs as paths
v0.3.5
    2012-07-25:
        * added optional path parameter to putData(); this is 
            useful to xlMFilter
v0.3.4
    2012-07-02:
        * added Server.exists(key) and Server.getData(key)

v0.3.3
    2012-07-01:
        * permit embedded slashes in uDir but drop any trailing 
            slashes
            - done for upaxUpdateNodeID and upaxBulkPoster
v0.3.2
    2012-06-02
        * corrected import errors hidden by the presence of an
            XLattice.Core.NodeID package; these errors were in 
            upax/ftlog.py and in the unit tests

v0.3.1
    2012-06-02
        * Added locking using u.ULock to upaxBulkPoster

upax/ftlog.py
# dev/py/upax3/upax3/ftlog.py

""" Fault-tolerant log for a Upax node. """

import os
import re
# import sys
from collections import Container, Sized
from xlattice import (HashTypes, check_hashtype,     # u,
                      SHA1_HEX_NONE, SHA2_HEX_NONE, SHA3_HEX_NONE)
from upax import UpaxError
from upax.node import check_hex_node_id_160, check_hex_node_id_256

__all__ = ['ATEXT', 'AT_FREE',
           'PATH_RE',
           'BODY_LINE_1_RE', 'BODY_LINE_256_RE',
           'IGNORABLE_RE',

           # classes
           'Log', 'BoundLog', 'LogEntry',
           'Reader', 'FileReader', 'StringReader', ]

# -------------------------------------------------------------------
# CLASS LOG AND SUBCLASSES
# -------------------------------------------------------------------

# Take care: this pattern is used in xlmfilter, possibly elsewhere
# this is RFC2822's atext; *,+,?,- are escaped; needs to be enclosed in []+
ATEXT = r"[a-z0-9!#$%&'\*\+/=\?^_`{|}~\-]+"
AT_FREE = ATEXT + r'(?:\.' + ATEXT + r')*'

# this permits an RFC2822 message ID but is a little less restrictive
PATH_PAT = AT_FREE + r'(?:@' + AT_FREE + ')?'
PATH_RE = re.compile(PATH_PAT, re.I)

BODY_LINE_1_PAT =\
    r'^(\d+) ([0-9a-f]{40}) ([0-9a-f]{40}) "([^"]*)" (%s)$' % PATH_PAT
BODY_LINE_1_RE = re.compile(BODY_LINE_1_PAT, re.I)

BODY_LINE_256_PAT =\
    r'^(\d+) ([0-9a-f]{64}) ([0-9a-f]{64}) "([^"]*)" (%s)$' % PATH_PAT
BODY_LINE_256_RE = re.compile(BODY_LINE_256_PAT, re.I)


IGNORABLE_PAT = '(^ *$)|^ *#'
IGNORABLE_RE = re.compile(IGNORABLE_PAT)


class Log(Container, Sized):
    """a fault-tolerant log"""

    def __init__(self, reader, hashtype):
        self._hashtype = hashtype
        (timestamp, prev_log_hash, prev_master, entries, index) = reader.read()
        self._timestamp = timestamp     # seconds from epoch
        self._prev_hash = prev_log_hash   # SHA1/3 hash of previous Log
        if hashtype == HashTypes.SHA1:
            check_hex_node_id_160(self._prev_hash)
        else:
            check_hex_node_id_256(self._prev_hash)
        self._prev_master = prev_master    # nodeID of master writing prev log
        if hashtype == HashTypes.SHA1:
            check_hex_node_id_160(self._prev_master)
        else:
            check_hex_node_id_256(self._prev_master)

        self._entries = entries       # a list
        self._index = index         # a map, hash => entry

    def __contains__(self, key):
        """ Return whether this key is in the Log. """
        return key in self._index

    def __len__(self):
        """ Return the length of this Log. """
        return len(self._entries)

    def __str__(self):
        """used for serialization, so includes newline"""

        # first line
        if self._hashtype == HashTypes.SHA1:
            fmt = "%013u %40s %40s\n"
        else:
            fmt = "%013u %64s %64s\n"
        ret = fmt % (self._timestamp, self._prev_hash, self._prev_master)

        # list of entries
        for entry in self._entries:
            ret += str(entry)           # woefully inefficient :-)
        return ret

    def add_entry(self, tstamp, key, node_id, src, path):
        """
        Create a LogEntry with the given timestamp, key, nodeID, src, and path.

        If the LogEntry is already present in the Log, return a reference to
        the existing LogEntry.  Otherwise, add the LogEntry to the list and
        index it by key.
        """
        entry = LogEntry(tstamp, key, node_id, src, path)
        if key in self._index:
            existing = self._index[key]
            if entry == existing:
                return existing         # silently ignore duplicates
        self._entries.append(entry)     # increases size of list
        self._index[key] = entry        # overwrites any earlier duplicates
        return entry

    def get_entry(self, key):
        """ Given a key, return the corresponding LogEntry or None. """
        if not key in self._index:
            return None
        else:
            return self._index[key]

    @property
    def entries(self):
        """ Return the list of LogEntries. """
        return self._entries

    @property
    def index(self):
        """ Return the index by key into the list of LogEntries. """
        return self._index

    @property
    def prev_hash(self):
        """ Return the content hash of the previous Log. """
        return self._prev_hash

    @property
    def prev_master(self):
        """
        Return the ID of the master of the previous Log.
        """
        return self._prev_master

    @property
    def timestamp(self):
        """ Return the timestamp for this Log. """
        return self._timestamp


class BoundLog(Log):
    """ A fult tolerant log bound to a file. """

    def __init__(self, reader, hashtype=HashTypes.SHA2,
                 u_path=None, base_name='L'):
        super(). __init__(reader, hashtype)
        self.fd_ = None
        self.is_open = False     # for appending
        overwriting = False
        if u_path:
            self.u_path = u_path
            self.base_name = base_name
            overwriting = True
        else:
            if isinstance(reader, FileReader):
                self.u_path = reader.u_path
                self.base_name = reader.base_name
                overwriting = False
            else:
                msg = "no target uPath/baseName specified"
                raise UpaxError(msg)
        self.path_to_log = "%s/%s" % (self.u_path, self.base_name)
        if overwriting:
            with open(self.path_to_log, 'w') as file:
                log_contents = super(BoundLog, self).__str__()
                file.write(log_contents)
                file.close()
        self.fd_ = open(self.path_to_log, 'a')
        self.is_open = True

    def add_entry(self, tstamp, key, node_id, src, path):
        if not self.is_open:
            msg = "log file %s is not open for appending" % self.path_to_log
            raise UpaxError(msg)

        # XXX NEED TO THINK ABOUT THE ORDER OF OPERATIONS HERE
        entry = super(
            BoundLog,
            self).add_entry(tstamp, key, node_id, src, path)
        stringified = str(entry)
        self.fd_.write(stringified)
        return entry

    def flush(self):
        """
        Flush the log.

        This should write the contents of any internal buffers to disk,
        but no particular behavior is guaranteed.
        """
        self.fd_.flush()

    def close(self):
        """ Close the log. """
        self.fd_.close()
        self.is_open = False


# -------------------------------------------------------------------
class LogEntry():
    """
    The entry made upon adding a file to the Upax content-keyed data store.

    This consists of a timestamp; an SHA content key, the hash of the
    contents of the file, the NodeID identifying the contributor,
    its source (which may be a program name, and a UNIX/POSIX path
    associated with the file.  The path will normally be relative.
    """
    __slots__ = ['_timestamp', '_key', '_node_id', '_src', '_path', ]

    def __init__(self,
                 timestamp, key, node_id, source, pathToDoc):
        self._timestamp = timestamp      # seconds from epoch

        if key is None:
            raise UpaxError('LogEntry key may not be None')
        hashtype = len(key) == 40
        self._key = key              # 40 or 64 hex digits, content hash
        if hashtype == HashTypes.SHA1:
            check_hex_node_id_160(self._key)
        else:
            check_hex_node_id_256(self._key)

        if node_id is None:
            raise UpaxError('LogEntry nodeID may not be None')
        self._node_id = node_id           # 40/64 digits, node providing entry
        # XXX This is questionable.  Why can't a node with a SHA1 id store
        # a datum with a SHA3 key?
        if hashtype == HashTypes.SHA1:
            check_hex_node_id_160(self._node_id)
        else:
            check_hex_node_id_256(self._node_id)

        self._src = source           # tool or person responsible
        self._path = pathToDoc        # file name

    @property
    def key(self):
        """
        Return the 40- or 64-byte SHA hash associated with the entry.

        This is an SHA content hash.
        """
        return self._key

    @property
    def node_id(self):
        """ Return the 40- or 64-byte NodeID associated with the entry. """
        return self._node_id

    @property
    def path(self):
        """ Return the POSIX path associated with the LogEntry. """
        return self._path

    @property
    def src(self):
        """ Return the 'src' associated with the LogEntry. """
        return self._src

    @property
    def timestamp(self):
        """ Return the time at which the LogEntry was created. """
        return self._timestamp

    @property
    def hashtype(self):
        """ XXX WRONG should return key length, allowing 64 or 40. """
        return len(self._key) == 40

    # used in serialization, so newlines are intended
    def __str__(self):
        if self.hashtype == HashTypes.SHA1:
            fmt = '%013u %40s %40s "%s" %s\n'
        else:
            fmt = '%013u %64s %64s "%s" %s\n'
        return fmt % (self._timestamp, self._key,
                      self._node_id, self._src, self._path)

    def __eq__(self, other):
        return isinstance(other, LogEntry) and\
            self._timestamp == other.timestamp and\
            self._key == other.key and\
            self._node_id == other.node_id and\
            self._src == other.src and\
            self._path == other.path

    def __ne__(self, other):
        return not self.__eq__(other)

    def equals(self, other):
        """
        The function usualy known as __eq__.  XXX DEPRECATED
        """
        return self.__eq__(other)

# -------------------------------------------------------------------
# CLASS READER AND SUBCLASSES
# -------------------------------------------------------------------


class Reader(object):
    """
    Would prefer to be able to handle this through something like a Java
    Reader, so that we could test with a StringReader but then use a
    FileReader in production.  If it is a file, file.readlines(sizeHint)
    supposedly has very good preformance for larger sizeHint, say 100KB
    It appears that lines returned need to be rstripped, which wastefully
    requires copying

    For our purposes, string input can just be split on newlines, which
    has the benefit of effectively chomping at the same time
    """

    #__slots__ = ['_entries', '_index', '_lines', '_hashtype',
    #             'FIRST_LINE_RE', ]

    def __init__(self, lines, hashtype):
        check_hashtype(hashtype)
        self._hashtype = hashtype
        if hashtype == HashTypes.SHA1:
            first_line_pat = r'^(\d{13}) ([0-9a-f]{40}) ([0-9a-f]{40})$'
        else:
            first_line_pat = r'^(\d{13}) ([0-9a-f]{64}) ([0-9a-f]{64})$'
        self.first_line_re = re.compile(first_line_pat, re.I)

        # XXX verify that argument is an array of strings
        self._lines = lines

        ndx_last = len(self._lines) - 1
        # strip newline from last line if present
        if ndx_last >= 1:
            self._lines[ndx_last] = self._lines[ndx_last].rstrip('\n')

        # Entries are a collection, a list.  We also need a dictionary
        # that accesses each log entry using its hash.
        self._entries = []            # the empty list
        self._index = dict()        # mapping hash => entry

    @property
    def hashtype(self):
        """ Return the type of SHA hash used. """
        return self._hashtype

    def read(self):
        """
        The first line contains timestamp, hash, nodeID for previous Log.
        Succeeding lines look like
           timestamp hash nodeID src path
        In both cases timestamp is an unsigned int, the number of
        milliseconds since the epoch.  It can be printed with %13u.
        The current value (April 2011) is about 1.3 trillion (1301961973000).
        """

        first_line = None
        if self._lines and len(self._lines) > 0:
            first_line = self._lines[0]
        if first_line:
            match = re.match(self.first_line_re, first_line)
            if not match:
                print("NO MATCH, FIRST LINE; hashtype = %s" % self.hashtype)
                print(("  FIRST LINE: '%s'" % first_line))
                raise UpaxError("no match on first line; giving up")
            timestamp = int(match.group(1))
            prev_log_hash = match.group(2)
            prev_master = match.group(3)
            del self._lines[0]        # so we can cleanly iterate
        else:
            # no first line
            timestamp = 0
            if self._hashtype == HashTypes.SHA1:
                prev_log_hash = SHA1_HEX_NONE
                prev_master = SHA1_HEX_NONE
            elif self._hashtype == HashTypes.SHA2:
                prev_log_hash = SHA2_HEX_NONE
                prev_master = SHA2_HEX_NONE
            elif self._hashtype == HashTypes.SHA3:
                prev_log_hash = SHA3_HEX_NONE
                prev_master = SHA3_HEX_NONE

        entries = []
        index = dict()

        for line in self._lines:
            # Read each successive line, creating an entry for each and
            # indexing each.  Ignore blank lines and those beginning with
            # a hash ('#')
            match = re.match(IGNORABLE_RE, line)
            if match:
                continue
            if self._hashtype == HashTypes.SHA1:
                match = re.match(BODY_LINE_1_RE, line)
            else:
                match = re.match(BODY_LINE_256_RE, line)
            if match:
                tstamp = int(match.group(1))
                key = match.group(2)
                node_id = match.group(3)
                src = match.group(4)
                path = match.group(5)
                # constructor should catch invalid fields
                entry = LogEntry(tstamp, key, node_id, src, path)
                entries.append(entry)
                index[key] = entry
            else:
                msg = "not a valid log entry line: '%s'" % line
                raise UpaxError(msg)

        return (timestamp, prev_log_hash, prev_master, entries, index)

# -------------------------------------------------------------------


class FileReader(Reader):
    """
    Accept uPath and optionally log file name, read entire file into
    a string array, pass to Reader.
    """
    __slots__ = ['_u_path', '_base_name', '_log_file', ]

    # XXX CHECK ORDER OF ARGUMENTS
    def __init__(self, u_path, hashtype=False, base_name="L"):
        if not os.path.exists(u_path):
            raise UpaxError("no such directory %s" % u_path)
        self._u_path = u_path
        self._base_name = base_name
        self._log_file = "%s/%s" % (self._u_path, base_name)
        with open(self._log_file, 'r') as file:
            contents = file.read()
        lines = contents.split('\n')
        super(FileReader, self).__init__(lines, hashtype)

    @property
    def base_name(self):
        """ Return the base name of the log file. """
        return self._base_name

    @property
    def log_file(self):
        """ Return the path to the log file. """
        return self._log_file

    @property
    def u_path(self):
        """ Return the path to uDir, the content-keyed store. """
        return self._u_path

# -------------------------------------------------------------------


class StringReader(Reader):
    """
    Accept a (big) string, convert to a string array, pass to Reader
    """

    def __init__(self, bigString, hashtype=False):

        # split on newlines
        lines = bigString.split('\n')

        super().__init__(lines, hashtype)
